### ISSUE TITLE
Improve Binary

### ISSUE THREAD

**Issue by Yoric** [fxbox/taxonomy#37]
For the moment, we handle Binary as a `Vec<u8>` passed through JSON. That's pretty ugly, so we should find a better way to store Binaries.
- One possibility would be to transmit them as a multipart request/response. I believe that `fetch` handles that nicely, but I haven't checked.
- Another possibility would be to expose them as randomly-generated, refcounted + short-lived URIs.

Cc-ing people who had to deal with Binary: @dhylands , @azasypkin .

**Comment by fabricedesre** [comment]
When a getter returns a single binary response, we actually turn it into a proper http response with the right mime type instead of the json serialization of the `Vec<u8>`.
Multi-part would be better when we have multiple Values for sure.
Short-lived URIs are an interesting idea, but how long should we keep them around? That's also one more roundtrip...

And we should also figure out something for binary data posted to setters.

**Comment by Yoric** [comment]
> When a getter returns a single binary response, we actually turn it into a proper http response with the right mime type instead of the json serialization of the Vec<u8>.

Ah, forgot that. But that won't scale up to handle the full `ResultMap<>`.

> Short-lived URIs are an interesting idea, but how long should we keep them around? That's also one more roundtrip...

I would say: keep them until refcount is 0 and either 10 minutes have elapsed or memory/disk pressure. Good point about the roundtrip.

> And we should also figure out something for binary data posted to setters.

Yes. For that, I believe that multipart is the best choice.

**Comment by Yoric** [comment]
Cc-ing @gmarty , who's the main producer/consumer of data through the REST API.

**Comment by dhylands** [comment]
Looking at rust file I/O it appears that when using fixed sized buffers, it
uses [u8] and when using a buffer which will have an unknown size, then
Vec<u8> is used.

So that seems natural for dealing with binary data.

Dave Hylands

On Wed, Mar 30, 2016 at 11:24 PM, David Rajchenbach-Teller <
notifications@github.com> wrote:

> Cc-ing @gmarty https://github.com/gmarty , who's the main
> producer/consumer of data through the REST API.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/fxbox/taxonomy/issues/37#issuecomment-203775726

## 

Dave Hylands
Shuswap, BC, Canada
http://www.davehylands.com

**Comment by azasypkin** [comment]
Indeed, from the Taxonomy side it feels correct to deal with `Vec<u8>/[u8]` (return or accept). It's up to API consumer (Foxbox) to (de)serialize it somehow if needed (e.g. "lazy" consumer can rely on (de)serialization mechanism provided by Taxonomy by default - JSON).

---- Foxbox part-----

In our case, multipart/mixed HTTP response for binary data looks like the optimal solution, so that _value_ response part will alternate with _error_ response part if any. The only my concern if it's supported well by our target browsers.

Also being a client side developer, I'd be happy to have separate endpoints for single binary blob (likely the most frequent use case) and for the list of blobs where single binary blob can be accessed via `GET` (instead of `PUT` that we use/abuse now) so that it can be directly used as `img[src]` or `a[download]` (instead of dealing with manually created/disposed object URLs).

**Comment by Yoric** [comment]
> Indeed, from the Taxonomy side it feels correct to deal with Vec<u8>/[u8](return or accept).

Agreed. I added an `Arc` to avoid having to duplicate the content whenever we duplicate a `Value`, which we sometimes do. It may be nicer to turn this into an `Arc<Value>` internally and be done with it.

> In our case, multipart/mixed HTTP response for binary data looks like the optimal solution, so that value response part will alternate with error response part if any. The only my concern if it's supported well by our target browsers.

We should check that out.

> Also being a client side developer, I'd be happy to have separate endpoints for single binary blob (likely the most frequent use case) and for the list of blobs where single binary blob can be accessed via GET (instead of PUT that we use/abuse now) so that it can be directly used as img[src] or a[download](instead of dealing with manually created/disposed object URLs).

I see two ways of doing this:
1. if you know the exact id of your getter, you could have an entry point `/api/v1/getter/$id/value`, but this doesn't fit too well with the fact that getters can have arbitrary ids (which may contain `../`, for instance), nor with selectors;
2. short-lived URLs, through which `Binary { value, mimetype }` is exported as `{"Binary": url}`.

I don't really like 1.

I'm ok with both 2. and multipart, or with having both and an option to pick either.

I believe that we should go ahead with multipart for the time being, as this is most likely the simplest to implement (i.e. we don't need to come up with a caching strategy), and then add an option to get a temporary URL. Or we can decide that providing both is too complex and concentrate on the multipart.

**Comment by fabricedesre** [comment]
This is not only about multi-part, but also the http method we use. `<img src>` and `<a download>` only send GET requests with no body, and we don't support that well.

**Comment by Yoric** [comment]
Nor do we really have a way to, as this is pretty much incompatible with selectors.

Unless we use short-lived URLs.

**Comment by Yoric** [comment]
Looking around the web, it seems that returning a multipart from the server is something that pretty much no web developer expects. So I'd go for short-lived URLs.

**Comment by Yoric** [comment]
Actually, I realize that this is a problem already faced by web developers everywhere. When you load a page, you typically need to load a number of images (or other resources). No web server that I know of returns a multipart. On the other hand, I believe that HTTP/2 web servers can return several resources in a single response, using a different mechanism.

Need to investigate this.

**Comment by Yoric** [comment]
After checking with @samgiles , it seems that HTTP/2 Server Push support is not available yet in Iron or Hyper. I'm now officially convinced that the way forward is temporary URLs – and we'll introduce Server Push as an optimization once it becomes available.
